yt/data_objects/static_output.py

```diff
@@ -1872,6 +1879,77 @@ def __eq__(self, other):
     def __hash__(self):
         return hash((self.filename, self.file_id, self.start, self.end))
 
+    def _nonzero_ptf(self, ptf):
+        for ptype, field_list in sorted(ptf.items()):
+            if self.total_particles[ptype] == 0:
+                continue
+            yield ptype, field_list, self.total_particles[ptype]
+
+    def _nonzero_ptypes(self, ptypes=None):
+        if ptypes is None:
+            ptypes = self.total_particles.keys()
+
+        for ptype in sorted(ptypes):
+            if self.total_particles[ptype] == 0:
+                continue
+            yield ptype, self.total_particles[ptype]
+
+    @contextlib.contextmanager
+    def transaction(self, handle=None):
+        # yields an open file handle, without double-reading.
+        if handle is None:
+            with self.open_handle() as handle:
+                yield handle
+        else:
+            # we have end up here recursively so simply yield and the outer
+            # transaction call will handle closing
+            yield handle
+
+    @abc.abstractmethod
+    @contextlib.contextmanager
+    def open_handle(self):
+        # this method is a file context manager that must yield an open file
+        # handle for self.filename.
+        pass
+
+    def _read_field(self, ptype: str, field: str, handle=None) -> Optional[np.ndarray]:
+        """this method wraps the actual data read (in _read_from_handle) in a
+        transaction context, preventing double reads.
+
+        Parameters
+        ----------
+        ptype : particle type
+        field : field name
+        handle : an optional already-open file handle
+
+        Returns
+        -------
+        np.ndarray or None : the data read from disk
+        """
+        with self.transaction(handle) as f:
+            data = self._read_from_handle(f, ptype, field)
+        return data
+
+    # @abc.abstractmethod
+    def _read_from_handle(self, handle, ptype: str, field: str) -> Optional[np.ndarray]:
+        """
+        This method must read data from an open file handle. It generally
+        should not be called directly except by self._read_field, but must be
+        implemented by each frontend. It can return None if no data is found.
+
+        Parameters
+        ----------
+        handle : an open file handle
+        ptype : particle type
+        field : field name
+
+        Returns
+        -------
+        np.ndarray or None (should be np array_like more generally?)
+        """
+        # TO DO: switch to an abstract method when existing frontends implement it
```

Comment:
doesn't seem like a lot but this is the dominant convention in the code base so it'd be a shame to make this one escape grepping :)
```suggestion
        # TODO: switch to an abstract method when existing frontends implement it
```

though what would really lower the chance that this goes forgotten would be to open an issue and link it in the comment